common/djangoapps/student/models.py

```diff
@@ -3116,6 +3120,130 @@ class AccountRecoveryConfiguration(ConfigurationModel):
     )
 
 
+class UserCelebration(TimeStampedModel):
+    """
+    Keeps track of how we've celebrated a user's progress on the platform.
+    This class is for course agnostic celebrations (not specific to a particular enrollment).
+    CourseEnrollmentCelebration is for celebrations that happen separately for each separate course.
+
+    The first_day_of_streak and last_day_of_streak fields are used to
+    control celebration of the streak feature.
+    A streak is when a learner visits the learning MFE N days in a row.
+    The business logic of streaks for a 3 day streak and 1 day break is the following:
+    1. Each streak should be celebrated exactly once, once the learner has completed the streak.
+    2. If a learner misses enough days to count as a break, the streak resets back to 0.
+    3. The streak is measured against the learner's configured timezone
+    4. We keep track of the total length of the streak, so there is a possibility in the future
+    to add multiple celebrations for longer streaks.
+    5. We keep track of the longest_ever_streak field for potential future use for badging purposes.
+
+    .. no_pii:
+    """
+    user = models.OneToOneField(User, models.CASCADE, related_name='celebration')
+    first_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    last_day_of_streak = models.DateField(default=None, null=True, blank=True)
+    longest_ever_streak = models.IntegerField(default=0)
+    STREAK_LENGTHS_TO_CELEBRATE = [3]
+    STREAK_BREAK_LENGTH = 1
+
+    def __str__(self):
+        return (
+            '[UserCelebration] user: {}; first_day_of_streak {};  last_day_of_streak {}; longest_ever_streak {};'
+        ).format(self.user.username, self.first_day_of_streak, self.last_day_of_streak, self.longest_ever_streak)
+
+    @classmethod
+    def _get_now(cls):
+        """ Retrieve the value for the current datetime in the user's timezone """
+        # importing here to avoid a circular import
+        from lms.djangoapps.courseware.context_processor import user_timezone_locale_prefs
+        user_timezone_locale = user_timezone_locale_prefs(crum.get_current_request())
+        user_timezone = timezone(user_timezone_locale['user_timezone'] or str(UTC))
+        return user_timezone.localize(datetime.now())
+
+    def _calculate_streak_updates(self):
+        """ Calculate the updates that should be applied to the streak fields of the provided celebration """
+        first_day_of_streak = self.first_day_of_streak
+        last_day_of_streak = self.last_day_of_streak
+        today = self.__class__._get_now().date()  # pylint: disable=protected-access
```

Comment:
You can call class methods on `self` directly: `self._get_now().date()`

Reply:
fixed